src/humanloop/eval_utils.py

```diff
@@ -0,0 +1,543 @@
+        If not provided your pipline must return a single string.
+    :param workers: the number of threads to process datapoints using your pipeline concurrently.
+    :return: per Evaluator checks.
+    """
+
+    # Get or create the file on Humanloop
+    version = file.pop("version", {})
+    try:
+        type_ = file.pop("type")
+    except KeyError as _:
+        # Default to flows if not type specified
+        type_ = "flow"
+        logger.warning("No type specified, defaulting to 'flow'.")
+
+    file_dict = {**file, **version}
+    match type_:
+        case "flow":
+            # Be more lenient with Flow versions as they are arbitrary json
+            try:
+                FlowKernel.parse_obj(version)
+            except ValidationError:
+                version = {"attributes": version}
+                file_dict = {**file, **version}
+            file = client.flows.upsert(**file_dict)
+        case "prompt":
+            file = client.flows.upsert(**file_dict)
+        case "tool":
+            file = client.flows.upsert(**file_dict)
+        case "evaluator":
+            file = client.evaluators.upsert(**file_dict)
+        case _:
+            raise NotImplementedError(f"Unsupported File type: {type_}")
+
+    # Upsert the Dataset
+    hl_dataset = client.datasets.upsert(**dataset)
+    hl_dataset = client.datasets.get(id=hl_dataset.id, include_datapoints=True)
+
+    # Upsert the local Evaluators; other Evaluators are just referenced by path
+    local_evaluators: list[RunEvaluator] = []
+    if evaluators:
+        for evaluator in evaluators:
+            # If a callable is provided for an Evaluator, we treat it as External
+            eval_callable = evaluator.get("callable")
+            if eval_callable is not None:
+                local_evaluators.append(evaluator)
+                spec = ExternalEvaluator(
+                    arguments_type=evaluator["args_type"],
+                    return_type=evaluator["return_type"],
+                    attributes={"code": inspect.getsource(eval_callable)},
+                    evaluator_type="external"
+                )
+                _ = client.evaluators.upsert(
+                    id=evaluator["id"],
+                    path=evaluator["path"],
+                    spec=spec
+                )
+    # Validate upfront that the local Evaluators and Dataset fit
+    requires_target = False
+    for local_evaluator in local_evaluators:
+        if local_evaluators["args_type"] == "target_required":
+            requires_target = True
+            break
+    if requires_target:
+        missing_target = 0
+        for datapoint in dataset.datapoints:
+            if not datapoint.target:
+                missing_target +=1
+        if missing_target > 0:
+            raise ValueError(f"Datapoint {datapoint.id} has no target is required for this Evaluator: {local_evaluator['path']}")
```

Comment:
Why not raise directly in the body of the for? datapoint will point to the latest point in the dataset, not the one missing the `target`

Reply:
Good spot. The message in the ValueError is just wrong - it should talk about the number of missings. I changed the logic and didn't change the error.